Coupling Relationship Managers

lindsayad opened the pull request:
- Removes libMesh default coupling functor, and instead explicitly requests the coupling needed from `KernelBase`, and `BoundaryCondition` (0 layers of ghosting)
- Ensures that correct sparsity is requested for `DGKernelBase` and `InterfaceKernelBase` simply by including the `RelationshipManagerType::COUPLING`
- When a RM of type ALGEBRAIC and COUPLING is requested, two relationship managers are created, one for the non-linear system to be added as a `coupling_functor` and another for the auxiliary system to be added as an `algebraic_ghosting_functor`. This is necessary because we want to ensure the correct `CouplingMatrix` is used for sparsity pattern generated in the non-linear system, and clearly that same `CouplingMatrix` has no meaning for the aux system

Closes #13736

lindsayad committed:
Add support for coupling functor relationship managers

Refs #13736

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/368861/) on 0347c43 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13750/site)

This comment will be updated on new commits.

lindsayad committed:
Do some RelationshipManager duplication

We want to have smart sparsity specification through our ghosting
functors. This means using our custom CouplingMatrices which
know the user-specified coupling between variables. This is
relevant for nonlinear-systems so we want to make sure we are
attaching that coupling information to ghosting functors operating
on the DofMap associated with the non-linear system. Obviously,
we cannot use that same coupling info for the aux system. There
it doesn't make any sense. Hence we must duplicate
RelationshipManagers; when the user requests an RM that is
evaluable and coupling, we must create two RMs: one that will
be attached to the non-linear DofMap that contains non-null
coupling matrices, and another that will be attached to the aux
DofMap that doesn't act on coupling information

lindsayad committed:
Increase coverage of no-mallocs tests

Now these tests will catch any time there are mallocs in
MatSetValues. This ensures that we're not getting mallocs
from the automatic scaling infrastructure or from the sparsity
pattern definition infrastructure created through our coupling
functors and other added sparsity functions

lindsayad committed:
Add no-mallocs test for kernels

This makes sure that we have added our default coupling functor
which set-ups the sparsity pattern for intra-element dof coupling

lindsayad committed:
Add description to geometric_edge_neighbors input file

lindsayad committed:
Really remove libMesh default ghosting/coupling

We are now fully explicit in the ghosting and sparsity coupling
pattern that we want. We add a coupling functor for kernels and
bcs that does not do any ghosting because it has _n_levels set
to 0, but it does specify intra-element dof coupling so that the
sparsity pattern is correct.

Now we can really re-gold the geometric edge test case because
there is really no evaluable ghosting now

lindsayad committed:
Remove mallocs from @WilkAndy test case

lindsayad committed:
Don't copy coupling functors between displaced and ref

We don't need to copy coupling back and forth between
ref and displaced because the displaced systems don't
hold their own matrices. And this is a really fortunate thing
because when computing the sparsity, libMesh passes one element
at a time from the active local elements to the coupling
functors. And in our ProxyRelationshipManager, every time
we call ProxyRelationshipManager::operator() we call through
to the other system's non-proxy coupling functors with ALL
the other mesh's active local elements. Consequently, there
is about a factor of 30 speed-up obtained by not copying coupling
functors, at least for the tensor_mechanics/examples/coal_mining.
cosserat_mc_wp_sticky_longitudinal_run test (which is only
a syntax check by the way)

permcody commented on framework/src/actions/Action.C:
```diff
@@ -89,102 +89,141 @@ Action::timedAct()
+        (rm_type != Moose::RelationshipManagerType::GEOMETRIC) &&
+        ((rm_type & Moose::RelationshipManagerType::GEOMETRIC) ==
+         Moose::RelationshipManagerType::GEOMETRIC) &&
+        rm_params.template get<bool>("attach_geometric_early"))))
```
Oh this is beautiful code...

permcody commented on framework/src/actions/Action.C:
```diff
@@ -89,102 +89,141 @@ Action::timedAct()
+    auto & rm_type = std::get<1>(buildable_type);
+    auto rm_input_parameter_func = std::get<2>(buildable_type);
+
+    // We need to be smart. If an object requested a relationship manager that is both an
```
> We need to be smart.

Hehe!

permcody reviewed (approved):
Looks really good! Thanks!
